File: leetcode/python/Trapping Rain Water/Trapping_Rain_Water.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def trap(self, A):
        left = 0
        right = len(A) - 1
        second_height = 0
        area = 0
        while(left < right):
            if A[left] < A[right]:
                second_height = max(second_height,A[left])
                area += second_height - A[left]
                left += 1
            else:
                second_height = max(second_height,A[right])
                area += second_height - A[right]
                right -= 1
        return area

File: leetcode/cpp/TrappingRainWater.cpp
//
// Created by zz on 2019/2/23.
//

// Time:  O(n)
// Space: O(1)

#include <vector>
using namespace std;



class Solution {
public:
    /**
     * @param heights: a list of integers
     * @return: a integer
     */
    int trapRainWater(vector<int> &heights) {
        if (heights.empty()){
            return 0;
        }
        int leftHeight = heights[0];
        int rightHeight = heights[heights.size() - 1];
        int trap = 0;
        int i = 0, j = heights.size() - 1;
        while (i < j){
            if (leftHeight < rightHeight){
                i++;
                leftHeight = max(leftHeight, heights[i]);
                trap += max(0, leftHeight - heights[i]);

            } else{
                j--;
                rightHeight = max(rightHeight, heights[j]);
                trap += max(0 ,rightHeight - heights[j]);

            }
        }
        return trap;

    }


};